lib/puppet/provider/package/pkg.rb

```diff
@@ -160,6 +160,10 @@ def insync?(is)
         warning(_("Implicit version %{should} has %{n} possible matches") % { should: should, n: n })
       end
       potential_matches.each { |p|
+        # If the version that we match is installed then we don't need
+        # to check if it's installed (or installable), just return true
+        return true if is != :absent && p[:status] == 'installed'
+
```

Comment:
To make sure I understand, suppose `git@2.15.2-11.4.0.0.1.9.0:20180618T171726Z` is installed and we specify a partial version like:

```
package { 'git':
  ensure => '2.15.2'
}
```

Prior to your PR, we'd call `pkg update -n git@<potential_version>` for each potential match, and stop as soon as the command returns 0 or 4.

Now with your PR, we'd stop if our desired version `2.15.2` is a prefix for any already installed version?

Reply:
Correct. Note that the the `pkg list` will order these so that we get the highest (most recent) version matching first.

Reply:
To be clear if there is a newer version that matches, that is not installed, it will get selected.

Reply:
> the pkg list will order these so that we get the highest (most recent) version matching first.

Does `pkg list` return the results in reverse natural order like:

```
git@2.100
git@2.50
git@2.15
```

Or does it return it in reverse lexicographic order?

```
git@2.50
git@2.15
git@2.100
```

It would be good to add a test that stubs the `pkg list` command to return multiple potential matches and then verify `expects(provider.insync?(is)).to eq(true)`